addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -3174,151 +3174,121 @@ export class OdooEditor extends EventTarget {
-                    const baseEmbedCommand = [
-                        {
-                            groupName: 'paste',
-                            title: this.options._t('Paste as URL'),
-                            description: this.options._t('Create an URL.'),
-                            fontawesome: 'fa-link',
-                            callback: () => {
-                                this.historyUndo();
-                                const link = document.createElement('A');
-                                link.setAttribute('href', url);
-                                for (const attribute in linkAttributes) {
-                                    link.setAttribute(attribute, linkAttributes[attribute]);
-                                }
-                                link.innerText = splitAroundUrl[i];
-                                const sel = this.document.getSelection();
-                                if (!sel.isCollapsed) {
-                                    this.deleteRange(sel);
-                                }
-                                if (sel.rangeCount) {
-                                    sel.getRangeAt(0).insertNode(link);
-                                    sel.collapseToEnd();
-                                }
-                            },
-                        },
-                        {
-                            groupName: 'paste',
-                            title: this.options._t('Paste as text'),
-                            description: this.options._t('Simple text paste.'),
-                            fontawesome: 'fa-font',
-                            callback: () => {},
-                        },
-                    ];
-
-                    const execCommandAtStepIndex = (index, callback) => {
-                        this._historyRevertUntil(index);
+                const isImageUrl = ['jpg', 'jpeg', 'png', 'gif', 'svg'].includes(urlFileExtention.toLowerCase());
+                // A url cannot be transformed inside an existing link.
+                // An image can be embedded inside an existing link, a video cannot.
+                if (selectionIsInsideALink) {
+                    if (isImageUrl) {
+                        const img = document.createElement('IMG');
+                        img.setAttribute('src', url);
+                        this._applyCommand('insertHTML', img);
+                    } else {
+                        this._applyCommand('insertText', text);
+                    }
+                } else if (isImageUrl || youtubeUrl) {
+                    // Open powerbox with commands to embed media or paste as link.
+                    // Insert URL as text, store history step index to revert it later.
+                    const stepIndexBeforeInsert = this._historySteps.length - 1;
+                    this._applyCommand('insertText', text);
+                    const execCommandAtStepIndex = (callback) => {
```

Comment:
1.
> The problem with removing the index paremeter is that now the name is somewhat wrong.

I agree. I could put it back, leaving this function as it was originally.

2.
> What do you think about the following rewrite ? Better or worse ?

Good, but not sure about the name. What we are actually reverting is the **text** insertion, and not "the powerbox". I would thus name it `revertTextInsertion`.

> I think it is easier to understand because the names are better, but to open the way for a better name I had to let go the automatic historyStep(true)

Actually, my bad here. I could go for `_applyCommand` here, or even keep it like this without the call do `historyStep`, as this will be called by Powebox' `postValidate` anyway.

3.
> This makes me think... what happens if you "open" the powerbox before inserting the text, then insert the text, then show the powerbox ? Wouldn't the default preValidate undo the text insertion just like we want without the need to manually do historyRevertUntil and all that ? (If it works, in 16.0 you could set this._beforeCommandbarStepIndex and this._isPowerboxOpenOnInput manually to achieve the same result.)

I would say that's a bit too much complexity just to avoid (re)defining a function. And we would have the challenge of dealing with the event handler put on keyup by Powebox.open:
```
const showOnceOnKeyup = () => {
            this.show();
            openOnKeyupTarget.removeEventListener('keyup', showOnceOnKeyup, true);
            initialTarget = openOnKeyupTarget;
            this._initialValue = openOnKeyupTarget.textContent;
        };
        openOnKeyupTarget.addEventListener('keyup', showOnceOnKeyup, true);
```
If you're willing to get hacky to avoid redefining a modified version of `preValidade`, we could leverage the fact that `shouldPrevalidade` is a function called right before calling the command's callback:
```
if (command) {
                !command.isIntermediateStep &&
                    (!command.shouldPreValidate || command.shouldPreValidate()) &&
                    this.options.preValidate &&
                    this.options.preValidate();
                command.callback();
                !command.isIntermediateStep &&
                    this.options.postValidate &&
                    this.options.postValidate();
            }
```

So we could do something like this:
```
const shouldPreValidate = () => {
   this._beforeCommandStepIndex = stepIndexBeforeInsert;
   return true;
}
```
And pass this on each command.
I don't lilke this actually, because you're producing a side effect with a function whose name would make you expect to return a boolean and do nothing else.

I prefer options 1 and 2 (with the proposed rename).
What do you think?

Reply:
Let's go for `revertTextInsertion` ! 👍